File: I semestr/ExecTimeComparison/decToBin.cpp
#include <iostream>
#include <chrono>
#include <string>

std::string decToBin(long long int decNumber) {
    std::string result{ "" };
    while (decNumber) {
        result += (decNumber % 2) + '0';
        decNumber /= 2;
    }

    return std::string(result.rbegin(), result.rend());
}

std::chrono::microseconds measure(long long count) {
    auto start = std::chrono::high_resolution_clock::now();

    for (unsigned long long i = 0; i < count; i++)
        decToBin(i);

    auto end = std::chrono::high_resolution_clock::now();
    return std::chrono::duration_cast<std::chrono::microseconds>(end - start);
}

int main() {
    long double measureSum{ 0 };
    long long int numOfCases{ 5000 };
    long count{ 1 };
    while (numOfCases < 1'000'000) {
        for (size_t i = 0; i < 100; i++)
            measureSum += measure(numOfCases).count();

        std::cout << "Average exec time for " << numOfCases << " cases: " << measureSum / 100 << " microseconds\n";
        numOfCases += 5000 * count;
        measureSum = 0;
        count += 1;
    }
}

File: I semestr/ExecTimeComparison/decToBin.rs
use std::time::{ Duration, Instant };

fn main() {
    let mut measure_sum: u128 = 0;
    let mut num_of_cases: i128 = 5000;
    let mut count: i64 = 1;

    while num_of_cases < 1000000 {
        for _i in 0..100 {
            measure_sum += measure(num_of_cases).as_micros()
        }

        println!("Average exec time for {} cases: {} mircoseconds", num_of_cases, measure_sum / 100);
        num_of_cases += 5000 * count as i128;
        measure_sum = 0;
        count += 1;
    }
}

fn dec_to_bin(mut dec_number: i128) -> String {
    let mut result = String::new();
    while dec_number > 0 {
        result.push({
            if dec_number % 2 == 0 {
                '0'
            } else {
                '1'
            }
        });

        dec_number /= 2;
    }


    result.chars().rev().collect()
}

fn measure(number_of_cases: i128) -> Duration {
    let start = Instant::now();

    for i in 0..number_of_cases {
        dec_to_bin(i);
    }

    start.elapsed()
}